### PR TITLE
Issue #47: Added custom AreaSummary entity and management reports

### PR DESCRIPTION
**Pull request by peacog**

### CODE REVIEW

File: web/modules/custom/area_summary/src/Entity/AreaSummary.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      ->setRevisionable(TRUE)
+      ->setDisplayOptions('form', ['weight' => 0]);
+
+    $fields['area_planted'] = BaseFieldDefinition::create('decimal')
+      ->setLabel(new TranslatableMarkup('Total area planted'))
+      ->setRevisionable(TRUE)
+      ->setDisplayOptions('form', ['weight' => 0]);
+
+    $fields['average_stems'] = BaseFieldDefinition::create('decimal')
+      ->setLabel(new TranslatableMarkup('Average stems/ha'))
```
Comment (pcambra) [inline web/modules/custom/area_summary/src/Entity/AreaSummary.php:103]: This is very interesting, I normally use a simple t()

File: web/modules/custom/area_summary/src/Form/AreaSummaryForm.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+   */
+  public function form(array $form, FormStateInterface $form_state) {
+    $form = parent::form($form, $form_state);
+    $entity = $this->entity;
+
+    // @todo: Is there a better way to check if an entity type is revisionable?
+    if ($entity->getEntityType()->hasKey('revision') && !$entity->isNew()) {
+      $form['revision'] = [
+        '#type' => 'checkbox',
+        '#title' => t('Create new revision'),
```
Comment (pcambra) [inline web/modules/custom/area_summary/src/Form/AreaSummaryForm.php:39]: use StringTranslationTrait; and then $this->t()

File: web/modules/custom/area_summary/src/Form/AreaSummaryForm.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ * Form controller for the area summary edit forms.
+ *
+ * @internal
+ */
+class AreaSummaryForm extends ContentEntityForm {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function prepareEntity() {
```
Comment (pcambra) [inline web/modules/custom/area_summary/src/Form/AreaSummaryForm.php:19]: Not sure if we need this?

File: web/modules/custom/area_summary/src/Form/AreaSummaryForm.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        );
+      }
+      else {
+        // Error on save.
+        $this->messenger()->addError($this->t('The entity could not be saved.'));
+        $form_state->setRebuild();
+      }
+    }
+    catch (\Exception $e) {
+      \Drupal::state()->set('entity_test.form.save.exception', get_class($e) . ': ' . $e->getMessage());
```
Comment (pcambra) [inline web/modules/custom/area_summary/src/Form/AreaSummaryForm.php:84]: Shouldn't be injected?

File: web/modules/custom/nfafmis/nfafmis.deploy.php [hunk 5]
```diff
@@ -199,3 +200,78 @@ function nfafmis_deploy_004(&$sandbox = NULL) {
     }
   }
 }
+
+/**
+ * Initialise CFR summary data by adding area_summary entities.
+ */
+function nfafmis_deploy_005(&$sandbox = NULL) {
```
Comment (pcambra) [inline web/modules/custom/nfafmis/nfafmis.deploy.php:207]: how does this deploy.php file works, is this a drush deploy thing?

File: web/modules/custom/area_summary/area_summary.module [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\node\Entity\Node;
+
+/**
+ * Implements hook_ENTITY_TYPE_update().
+ */
+function area_summary_node_update(EntityInterface $entity) {
+  if ($entity->bundle() == 'offer_license') {
+    // If a summary already exists for the area update it, otherwise create one.
+    $results = Drupal::entityQuery('area_summary')
```
Comment (pcambra) [inline web/modules/custom/area_summary/area_summary.module:18]: This should use entityTypeManager -> getStorage -> getQuery instead

File: web/modules/custom/area_summary/area_summary.module [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+          ->load($id);
+        $summary->area_allocated = $entity->field_overall_area->value;
+        $summary->save();
+      }
+    }
+    else {
+      $farmer = $entity->field_farmer_name_ref->entity;
+      $cfr = $entity->field_central_forest_reserve->entity;
+      $management_unit = $cfr->management_unit->entity;
+      $summary = AreaSummary::create([
```
Comment (pcambra) [inline web/modules/custom/area_summary/area_summary.module:36]: Same as above entityTypeManager -> getStorage ->create

File: web/modules/custom/area_summary/area_summary.module [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  }
+  elseif ($entity->bundle() == 'sub_area') {
+    // Recalculate the area planted total after the sub area was deleted.
+    $area = $entity->field_areas_id->entity;
+    _area_summary_update_area_planted($area);
+  }
+}
+
+/**
+ * Calculate
```
Comment (pcambra) [inline web/modules/custom/area_summary/area_summary.module:76]: Comments should end in a period, but we should add a more verbose description

File: web/modules/custom/area_summary/area_summary.module [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+ */
+function _area_summary_update_area_planted(EntityInterface $area) {
```
Comment (pcambra) [inline web/modules/custom/area_summary/area_summary.module:78]: I don't particularly recommend _prefixed functions, I tend to call them whatever_helper instead, it doesn't mean much anymore now that we have barely no hooks.

File: web/modules/custom/area_summary/area_summary.module [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  // Calculate the total area planted in sub-areas.
+  $sub_nids = Drupal::entityQuery('node')
```
Comment (pcambra) [inline web/modules/custom/area_summary/area_summary.module:80]: Same as above, it should be \Drupal::entityTypeManager()->getStorage('node')->getQuery()

File: web/modules/custom/area_summary/area_summary.module [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+  // A summary should already exist for the area so update it.
+  $results = Drupal::entityQuery('area_summary')
+    ->condition('area', $area->id())
+    ->accessCheck(FALSE)
+    ->execute();
+
+  foreach ($results as $id) {
+    $summary = \Drupal::entityTypeManager()->getStorage('area_summary')->load($id);
+    $summary->area_planted = $area_planted;
```
Comment (pcambra) [inline web/modules/custom/area_summary/area_summary.module:100]: If "area_planted" is a field, a ->set() method should be used instead.

File: web/modules/custom/area_summary/src/Controller/AreaSummaryController.php [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+class AreaSummaryController extends ControllerBase {
+
+  /**
+   * Builds the response.
+   */
+  public function build() {
+
+    $build['content'] = [
+      '#type' => 'item',
+      '#markup' => $this->t('It works!'),
```
Comment (pcambra) [inline web/modules/custom/area_summary/src/Controller/AreaSummaryController.php:19]: Is this on purpose?

File: web/modules/custom/area_summary/src/Controller/AreaSummaryListBuilder.php [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+ * @see \Drupal\nfafmis\Entity\AreaSummary
+ */
+class AreaSummaryListBuilder extends EntityListBuilder {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildHeader() {
+    $header['management_unit'] = t('Management unit');
+    $header['cfr'] = t('CFR');
```
Comment (pcambra) [inline web/modules/custom/area_summary/src/Controller/AreaSummaryListBuilder.php:20]: Can we use the translation trait here too and $this->t()?

File: web/modules/custom/area_summary/src/Controller/AreaSummaryListBuilder.php [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildRow(EntityInterface $entity) {
+    $row['management_unit'] = isset($entity->management_unit->entity) ? $entity->management_unit->entity->label() : NULL;
+    $row['cfr'] = $entity->cfr->entity->label();
+    $row['area'] = $entity->area->entity->label();
+    $row['area_allocated'] = $entity->area_allocated->value . ' ha';
```
Comment (pcambra) [inline web/modules/custom/area_summary/src/Controller/AreaSummaryListBuilder.php:35]: We normally use $entity->get('field_name')->value

File: web/modules/custom/nfafmis/nfafmis.deploy.php [hunk 15]
```diff
@@ -199,3 +200,78 @@ function nfafmis_deploy_004(&$sandbox = NULL) {
...
+  }
+
+  $batch_size = Settings::get('entity_update_batch_size', 50);
+  if (!empty($sandbox['nodes'])) {
+    // Handle nodes in batches.
+    $nids = array_slice($sandbox['nodes'], $sandbox['progress'], $batch_size);
+
+    $area_storage = \Drupal::entityTypeManager()->getStorage('node')->loadMultiple($nids);
+    foreach ($area_storage as $area) {
+      $farmer = $area->field_farmer_name_ref->entity;
```
Comment (pcambra) [inline web/modules/custom/nfafmis/nfafmis.deploy.php:234]: Same as above, ->get('field_name') instead

File: web/modules/custom/nfafmis/nfafmis.deploy.php (hunk 5 above)
Comment (peacog) [reply web/modules/custom/nfafmis/nfafmis.deploy.php:207]: Yes, it's for deploy hooks that are run after config has been imported. HOOK_post_update is called before drush cim and HOOK_deploy is called after drush cim.

File: web/modules/custom/area_summary/src/Entity/AreaSummary.php (hunk 1 above)
Comment (peacog) [reply web/modules/custom/area_summary/src/Entity/AreaSummary.php:103]: Simple t() is better :)

File: web/modules/custom/area_summary/src/Form/AreaSummaryForm.php (hunk 3 above)
Comment (peacog) [reply web/modules/custom/area_summary/src/Form/AreaSummaryForm.php]: We don't actually need the entity forms at all so I removed them to simplify things

File: web/modules/custom/area_summary/area_summary.module (hunk 11 above)
Comment (peacog) [reply web/modules/custom/area_summary/area_summary.module]: Done

File: web/modules/custom/area_summary/src/Controller/AreaSummaryController.php (hunk 12 above)
Comment (peacog) [reply web/modules/custom/area_summary/src/Controller/AreaSummaryController.php]: The controller wasn't needed to I've deleted it

**Pull request merged**